From 000c3951963b8ca8c068f425b6422ba7edd486e3 Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Sun, 15 Jan 2017 12:14:51 +0200
Subject: [PATCH 0759/1345] fix: net: mvpp2x: clear MIB statistic during port
 probe

commit  e9e3575bf9e9b4bad8d02233ac2747cfd7d30f5d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

Issue:
- MIB counters statistic wasn't cleared during probe and
  statistic from u-boot(TFTP and etc.) was counted.

Fix:
- Clear MIB statistic during port probe.

Change-Id: I08f082cbeeb0f0532d9f0e1ebe8d932a71e019ec
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35561
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c |   26 ++++++++++++++++++++
 drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.h |    1 +
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c |    3 +++
 3 files changed, 30 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
index 095e9c1..94488a2 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.c
@@ -2559,6 +2559,32 @@ void mv_gop110_mib_counters_stat_update(struct gop_hw *gop, int port, struct gop
 							MV_MIB_LATE_COLLISION);
 }
 
+void mv_gop110_mib_counters_clear(struct gop_hw *gop, int port)
+{
+	mv_gop110_mib_read64(gop, port, MV_MIB_GOOD_OCTETS_RECEIVED_LOW);
+	mv_gop110_mib_read64(gop, port, MV_MIB_UNICAST_FRAMES_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_BROADCAST_FRAMES_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_MULTICAST_FRAMES_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_GOOD_OCTETS_SENT_LOW);
+	mv_gop110_mib_read64(gop, port, MV_MIB_UNICAST_FRAMES_SENT);
+	mv_gop110_mib_read64(gop, port, MV_MIB_MULTICAST_FRAMES_SENT);
+	mv_gop110_mib_read64(gop, port, MV_MIB_BROADCAST_FRAMES_SENT);
+	mv_gop110_mib_read64(gop, port, MV_MIB_CRC_ERRORS_SENT);
+	mv_gop110_mib_read64(gop, port, MV_MIB_FC_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_FC_SENT);
+	mv_gop110_mib_read64(gop, port, MV_MIB_RX_FIFO_OVERRUN);
+	mv_gop110_mib_read64(gop, port, MV_MIB_UNDERSIZE_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_FRAGMENTS_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_OVERSIZE_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_JABBER_RECEIVED);
+	mv_gop110_mib_read64(gop, port, MV_MIB_MAC_RECEIVE_ERROR);
+	mv_gop110_mib_read64(gop, port, MV_MIB_BAD_CRC_EVENT);
+	mv_gop110_mib_read64(gop, port, MV_MIB_COLLISION);
+
+	/* This counter must be read last. Read it clear all read counters. */
+	mv_gop110_mib_read64(gop, port, MV_MIB_LATE_COLLISION);
+}
+
 void mv_gop110_netc_active_port(struct gop_hw *gop, u32 port, u32 val)
 {
 	u32 reg;
diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.h b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.h
index d77be0e..985cbb2 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.h
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_gop110_hw.h
@@ -433,6 +433,7 @@ static inline void mv_gop110_rfu1_print(struct gop_hw *gop, char *reg_name,
 void mv_gop110_mib_counters_show(struct gop_hw *gop, int port);
 void mv_gop110_mib_counters_stat_update(struct gop_hw *gop, int port,
 					struct gop_stat *gop_statistics);
+void mv_gop110_mib_counters_clear(struct gop_hw *gop, int port);
 
 /* PTP Functions */
 void mv_gop110_ptp_enable(struct gop_hw *gop, int port, bool state);
diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index f8bfebc..214e540 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -4701,6 +4701,9 @@ static int mv_pp2x_port_probe(struct platform_device *pdev,
 		goto err_free_port_pcpu;
 	}
 
+	/* Clear MIB counters statistic */
+	mv_gop110_mib_counters_clear(&port->priv->hw.gop, port->mac_data.gop_index);
+
 	mv_pp2x_port_irq_names_update(port);
 
 	netdev_info(dev, "Using %s mac address %pM\n", mac_from, dev->dev_addr);
-- 
1.7.9.5

